### PR TITLE
FWK-542 Add Azure Blob support

### PR DESCRIPTION
**Pull request by filkeith**

### CODE REVIEW

File: io/aws/s3/reader.go [hunk 1]
```diff
@@ -210,7 +209,7 @@ func (r *Reader) streamFile(
 		Key:    &filename,
 	})
 	if err != nil {
-		errorsCh <- fmt.Errorf("failed to open %s: %w", filename, err)
+		errorsCh <- fmt.Errorf("failerd to create reader from file %s: %w", filename, err)
```
Comment (reugn) [inline io/aws/s3/reader.go:212]:
```suggestion
		errorsCh <- fmt.Errorf("failed to create reader from file %s: %w", filename, err)
```

File: io/azure/blob/writer.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = <-w.done
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (reugn) [inline io/azure/blob/writer.go:185]:
```suggestion
	return <-w.done
```

File: io/azure/blob/writer.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blockblob"
+)
+
+const (
+	uploadStreamFileType    = "application/octet-stream"
+	uploadStreamBlockSize   = 5 * 1024 * 1024 // 5MB, minimum size of a part
+	uploadStreamConcurrency = 5
```
Comment (reugn) [inline io/azure/blob/writer.go:32]: It's better to make it an `option` and use this value by default.

File: io/azure/blob/writer.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+func newBlobWriter(ctx context.Context, blobClient *blockblob.Client) io.WriteCloser {
+	pipeReader, pipeWriter := io.Pipe()
+
+	w := &blobWriter{
+		blobClient: blobClient,
+		ctx:        ctx,
+		pipeReader: pipeReader,
+		pipeWriter: pipeWriter,
+		done:       make(chan error),
```
Comment (reugn) [inline io/azure/blob/writer.go:148]:
```suggestion
		done:       make(chan error, 1),
```

File: io/azure/blob/writer.go [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	go w.uploadStream()
+
+	return w
+}
+
+func (w *blobWriter) uploadStream() {
+	defer w.pipeReader.Close()
```
Comment (reugn) [inline io/azure/blob/writer.go:157]: Unhandled error on Close.

File: io/azure/blob/writer.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+func (w *Writer) NewWriter(ctx context.Context, filename string) (io.WriteCloser, error) {
+	// protection for single file backup.
+	if !w.isDir {
+		if !w.called.CompareAndSwap(false, true) {
+			return nil, fmt.Errorf("parallel running for single file is not allowed")
+		}
+	}
+	// If we use backup to single file, we overwrite the file name.
+	if !w.isDir {
+		filename = w.path
```
Comment (reugn) [inline io/azure/blob/writer.go:120]: Can be merged with the previous if.

File: io/azure/blob/writer.go (hunk 5 above)
Comment (filkeith) [reply io/azure/blob/writer.go]: it will be closed by backup, so removed this line

File: io/azure/blob/writer.go [hunk 7]
```diff
@@ -154,12 +162,10 @@ func newBlobWriter(ctx context.Context, blobClient *blockblob.Client) io.WriteCl
 }
 
 func (w *blobWriter) uploadStream() {
-	defer w.pipeReader.Close()
-
 	contentType := uploadStreamFileType
 	_, err := w.blobClient.UploadStream(w.ctx, w.pipeReader, &azblob.UploadStreamOptions{
 		BlockSize:   uploadStreamBlockSize,
-		Concurrency: uploadStreamConcurrency,
+		Concurrency: uploadStreamConcurrencyDefault,
```
Comment (reugn) [inline io/azure/blob/writer.go:168]: Take the value from options.

Comment (filkeith) [reply io/azure/blob/writer.go]: fixed

**Review by reugn — APPROVED** [review]

**Pull request merged**